### ISSUE TITLE
torch.tensor not copying to correct device if data is a Tensor with correct dtype

### ISSUE THREAD

**Issue by SsnL** [pytorch/pytorch#7277]
`torch.tensor` not respecting `device` when input is already a tensor of correct `dtype` (but of wrong `device`).

```
In [1]: import torch

In [2]: device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

In [3]: rand_input = torch.tensor(data=torch.Tensor(1, 3, 80, 170), device=device, requires_grad=True)

In [4]: rand_input.device
Out[4]: device(type='cpu')

In [5]: device
Out[5]: device(type='cuda')
```

reported at https://discuss.pytorch.org/t/torch-tensor-does-not-return-a-tensor-with-device-type-cuda/17436

cc @gchanan

**Comment by Arkoprabho** [comment]
Thank you for opening an issue here. I think it'll be better if the further conversation is carried out here.

**Comment by apaszke** [comment]
I think that's expected, since `torch.tensor` is said to always copy the data, and as such creates a brand new tensor. You should probably use `torch.as_tensor` in this case.

**Comment by fmassa** [comment]
I think if we specify the device explicitly, it should perform the move to the device. If no device is passed in the argument, then it should take the device of the tensor that was passed.

```python
a = torch.rand(3, device='cuda')

# what should be the device here?
# I'd say on the gpu
b = torch.tensor(a)

# but this should be on the cpu
c = torch.tensor(a, device='cpu')
```

**Comment by apaszke** [comment]
I don't think I agree. It's not `tensor_like`, it's a first class factory, that treats its argument as a "data source". All parameters are set to user defined defaults, unless overriden.

**Comment by fmassa** [comment]
Hum, I think this can get confusing.
Don't we have type inference for `torch.tensor`? Why can't we then have a `device` inference as well?

If `torch.as_tensor` does device inference as well, then I'm ok, but we really need an option that creates a tensor and infers the device from the data that was sent.

**Comment by apaszke** [comment]
I think both ways are confusing tbh 😄 In reality, you probably just should never use `torch.tensor`, because you rarely want to enforce the copy, even if the data is perfectly valid. Maybe let's see what @gchanan thinks?

**Comment by SsnL** [comment]
@apaszke Since it always copies `data`, shouldn't it just copy to the specified device? I can't really understand why this might be expected. Why does it even use the `data`'s device when a `device` arg is explicitly given? 

You also wrote 
> All parameters are set to user defined defaults, unless overriden.

Here the `device` is explicitly overridden.

**Comment by apaszke** [comment]
@SsnL I wasn't referring to the code in the issue (which is obviously broken by a bug), but to the comment of @fmassa:

> If no device is passed in the argument, then it should take the device of the tensor that was passed.

**Comment by zou3519** [comment]
cc @gchanan what do you think?

**Comment by shianiawhite** [comment]
Note, this also occurs when creating the array from a NumPy array (that took me a while to figure out what was causing the problem). I would definitely expect a non-PyTorch datatype to take the device setting into account:

```
import torch
import numpy as np

x = torch.tensor(np.array(1), device='cuda:0')
print(x.device)  # Prints `cpu`
x = torch.tensor(1, device='cuda:0')
print(x.device)  # Prints `cuda:0`
```

**Comment by gchanan** [comment]
The original issue and the one mentioned by @shianiawhite have been fixed on master.

**Comment by gchanan** [comment]
As to the issue @fmassa and @apaszke are discussing about whether the device should be inferred for `torch.tensor`...I think I agree with @apaszke that either choice is confusing :), but I think inferring the device is probably the least surprising.

Also, since we now do device inference (as noted above, the issues reported have been fixed), if the input is a tensor, `torch.tensor` behaves identically to `Tensor.new_tensor` (`torch.tensor` supports non-tensor inputs as well).  This is pretty strange, so maybe we should just get rid of `Tensor.new_tensor`.

> In reality, you probably just should never use torch.tensor, because you rarely want to enforce the copy, even if the data is perfectly valid. Maybe let's see what @gchanan thinks?

Somewhat interestingly, `numpy.array` has a `copy` parameter (default True), so you can avoid the copy, although in most cases it just seems easier to call `numpy.as_array` (except it lacks some parameters).  Maybe we should add that parameter, because what we have now is subtle (i.e. people will just call `torch.tensor` because the name makes more sense, and it's not obvious unless you look at the docs that what you want is `torch.as_tensor`).

**Comment by apaszke** [comment]
We should probably emphasise the tensor vs as_tensor difference somewhere at the beginning of our tutorials.

**Comment by Balandat** [comment]
This happens even when using a pure python list literal:

```
import torch

v1 = torch.randn(5).cuda()
v2 = torch.tensor([1.], device=v1.device)
v3 = torch.Tensor([1.], device=v1.device)

for v in (v1, v2, v3):
    print(v.device)

cuda:0
cuda:0
cpu
```

**Comment by heitorschueroff** [comment]
Issue was fixed and as suggested, torch.tensor takes the device of the tensor passed in. However, we discourage copying tensors using torch.tensor and prefer to use clone().detach()

**Comment by lyhyl** [comment]
> Issue was fixed and as suggested, torch.tensor takes the device of the tensor passed in. However, we discourage copying tensors using torch.tensor and prefer to use clone().detach()

![image](https://user-images.githubusercontent.com/3370369/101347167-9b965c00-38c4-11eb-8cd3-b5d5b3551964.png)
no, it still happened.
